File: internal/domain/report.go
package domain

import (
	"context"
	"fmt"
	"time"

	"github.com/leighmacdonald/steamid/v4/steamid"
)

type ReportRepository interface {
	GetReportBySteamID(ctx context.Context, authorID steamid.SteamID, steamID steamid.SteamID) (Report, error)
	GetReports(ctx context.Context, opts ReportQueryFilter) ([]Report, error)
	GetReport(ctx context.Context, reportID int64) (Report, error)
	GetReportMessages(ctx context.Context, reportID int64) ([]ReportMessage, error)
	GetReportMessageByID(ctx context.Context, reportMessageID int64) (ReportMessage, error)
	DropReportMessage(ctx context.Context, message *ReportMessage) error
	DropReport(ctx context.Context, report *Report) error
	SaveReport(ctx context.Context, report *Report) error
	SaveReportMessage(ctx context.Context, message *ReportMessage) error
}

type ReportUsecase interface {
	GetReportBySteamID(ctx context.Context, authorID steamid.SteamID, steamID steamid.SteamID) (Report, error)
	GetReports(ctx context.Context, user PersonInfo, opts ReportQueryFilter) ([]ReportWithAuthor, error)
	GetReport(ctx context.Context, curUser PersonInfo, reportID int64) (ReportWithAuthor, error)
	GetReportMessages(ctx context.Context, reportID int64) ([]ReportMessage, error)
	GetReportMessageByID(ctx context.Context, reportMessageID int64) (ReportMessage, error)
	DropReportMessage(ctx context.Context, message *ReportMessage) error
	DropReport(ctx context.Context, report *Report) error
	SaveReport(ctx context.Context, report *Report) error
	SaveReportMessage(ctx context.Context, message *ReportMessage) error
}

type CreateReportReq struct {
	SourceID        steamid.SteamID `json:"source_id"`
	TargetID        steamid.SteamID `json:"target_id"`
	Description     string          `json:"description"`
	Reason          Reason          `json:"reason"`
	ReasonText      string          `json:"reason_text"`
	DemoID          int64           `json:"demo_id"`
	DemoTick        int             `json:"demo_tick"`
	PersonMessageID int64           `json:"person_message_id"`
}

type ReportStatus int

const (
	AnyStatus                        = -1
	Opened              ReportStatus = 0
	NeedMoreInfo                     = 1
	ClosedWithoutAction              = 2
	ClosedWithAction                 = 3
)

func (status ReportStatus) String() string {
	switch status {
	case ClosedWithoutAction:
		return "Closed without action"
	case ClosedWithAction:
		return "Closed with action"
	case Opened:
		return "Opened"
	default:
		return "Need more information"
	}
}

type Report struct {
	ReportID        int64           `json:"report_id"`
	SourceID        steamid.SteamID `json:"source_id"`
	TargetID        steamid.SteamID `json:"target_id"`
	Description     string          `json:"description"`
	ReportStatus    ReportStatus    `json:"report_status"`
	Reason          Reason          `json:"reason"`
	ReasonText      string          `json:"reason_text"`
	Deleted         bool            `json:"deleted"`
	DemoTick        int             `json:"demo_tick"`
	DemoID          int64           `json:"demo_id"`
	PersonMessageID int64           `json:"person_message_id"`
	CreatedOn       time.Time       `json:"created_on"`
	UpdatedOn       time.Time       `json:"updated_on"`
}

func (report Report) Path() string {
	return fmt.Sprintf("/report/%d", report.ReportID)
}

func NewReport() Report {
	return Report{
		ReportID:     0,
		SourceID:     steamid.SteamID{},
		Description:  "",
		ReportStatus: 0,
		CreatedOn:    time.Now(),
		UpdatedOn:    time.Now(),
		DemoTick:     -1,
		DemoID:       0,
	}
}

type ReportWithAuthor struct {
	Author  Person   `json:"author"`
	Subject Person   `json:"subject"`
	Demo    DemoFile `json:"demo"`
	Report
}

type ReportMessage struct {
	ReportID        int64           `json:"report_id"`
	ReportMessageID int64           `json:"report_message_id"`
	AuthorID        steamid.SteamID `json:"author_id"`
	MessageMD       string          `json:"message_md"`
	Deleted         bool            `json:"deleted"`
	CreatedOn       time.Time       `json:"created_on"`
	UpdatedOn       time.Time       `json:"updated_on"`
	SimplePerson
}

func NewReportMessage(reportID int64, authorID steamid.SteamID, messageMD string) ReportMessage {
	now := time.Now()

	return ReportMessage{
		ReportID:     reportID,
		AuthorID:     authorID,
		MessageMD:    messageMD,
		Deleted:      false,
		CreatedOn:    now,
		UpdatedOn:    now,
		SimplePerson: SimplePerson{},
	}
}

type ReportMeta struct {
	TotalOpen   int
	TotalClosed int
	Open        int
	NeedInfo    int
	Open1Day    int
	Open3Days   int
	OpenWeek    int
	OpenNew     int
}

File: frontend/src/schema/report.ts
import { Theme } from '@mui/material';
import { z } from 'zod';
import { BanReasonEnum } from './bans.ts';
import { schemaTimeStamped } from './chrono.ts';
import { schemaDemoFile } from './demo.ts';
import { schemaPerson } from './people.ts';

export const ReportStatus = {
    Any: -1,
    Opened: 0,
    NeedMoreInfo: 1,
    ClosedWithoutAction: 2,
    ClosedWithAction: 3
} as const;

export const ReportStatusEnum = z.nativeEnum(ReportStatus);
export type ReportStatusEnum = z.infer<typeof ReportStatusEnum>;

export const ReportStatusCollection = [
    ReportStatus.Any,
    ReportStatus.Opened,
    ReportStatus.NeedMoreInfo,
    ReportStatus.ClosedWithoutAction,
    ReportStatus.ClosedWithAction
];

export const reportStatusString = (rs: ReportStatusEnum): string => {
    switch (rs) {
        case ReportStatus.NeedMoreInfo:
            return 'Need More Info';
        case ReportStatus.ClosedWithoutAction:
            return 'Closed Without Action';
        case ReportStatus.ClosedWithAction:
            return 'Closed With Action';
        case ReportStatus.Opened:
            return 'Opened';
        case ReportStatus.Any:
            return 'Any';
    }
};

export const reportStatusColour = (rs: ReportStatusEnum, theme: Theme): string => {
    switch (rs) {
        case ReportStatus.NeedMoreInfo:
            return theme.palette.warning.main;
        case ReportStatus.ClosedWithoutAction:
            return theme.palette.error.main;
        case ReportStatus.ClosedWithAction:
            return theme.palette.success.main;
        default:
            return theme.palette.info.main;
    }
};

export const schemaReport = z
    .object({
        report_id: z.number(),
        source_id: z.string(),
        target_id: z.string(),
        description: z.string(),
        report_status: ReportStatusEnum,
        deleted: z.boolean(),
        reason: BanReasonEnum,
        reason_text: z.string(),
        demo_name: z.string(),
        demo_tick: z.number(),
        demo_id: z.number(),
        person_message_id: z.number()
    })
    .merge(schemaTimeStamped);

export type Report = z.infer<typeof schemaReport>;

export const schemaBasicUserInfo = z.object({
    steam_id: z.string(),
    personaname: z.string(),
    avatarhash: z.string()
});
export type BasicUserInfo = z.infer<typeof schemaBasicUserInfo>;

export const schemaBanAppealMessage = z
    .object({
        ban_id: z.number(),
        ban_message_id: z.number(),
        author_id: z.string(),
        message_md: z.string(),
        deleted: z.boolean()
    })
    .merge(schemaTimeStamped)
    .merge(schemaBasicUserInfo);

export type BanAppealMessage = z.infer<typeof schemaBanAppealMessage>;

export const schemaReportMessage = z
    .object({
        report_id: z.number(),
        report_message_id: z.number(),
        author_id: z.string(),
        message_md: z.string(),
        deleted: z.boolean()
    })
    .merge(schemaTimeStamped)
    .merge(schemaBasicUserInfo);

export type ReportMessage = z.infer<typeof schemaReportMessage>;

export const schemaCreateReportRequest = z.object({
    target_id: z.string(),
    description: z.string().min(10),
    reason: BanReasonEnum,
    reason_text: z.string(),
    demo_id: z.number(),
    demo_tick: z.number(),
    person_message_id: z.number()
});

export type CreateReportRequest = z.infer<typeof schemaCreateReportRequest>;

export const schemaReportWithAuthor = z
    .object({
        author: schemaPerson,
        subject: schemaPerson,
        demo: schemaDemoFile
    })
    .merge(schemaReport);

export type ReportWithAuthor = z.infer<typeof schemaReportWithAuthor>;

export const schemaCreateReportMessage = z.object({
    body_md: z.string()
});

export type CreateReportMessage = z.infer<typeof schemaCreateReportMessage>;
